src/lib/middleware/pageCacheMiddleware.ts

```diff
@@ -0,0 +1,73 @@
+import {
+  // @ts-ignore
+  PAGE_CACHE_ENABLED,
+  // @ts-ignore
+  PAGE_CACHE_TYPES,
+  // @ts-ignore
+  PAGE_CACHE_NAMESPACE,
+  // @ts-ignore
+  PAGE_CACHE_VERSION,
+  // @ts-ignore
+  PAGE_CACHE_EXPIRY_SECONDS,
+} from "../../config"
+import { Request, Response, NextFunction } from "express"
+const cache = require("lib/cache.coffee")
+const cacheablePageTypes: string[] = PAGE_CACHE_TYPES.split("|")
+
+// Middleware will `next` and do nothing if any of the following is true:
+//
+// * page cache feature is disabled.
+// * a user is signed in.
+// * this isnt a supported cacheable path (there is an allow-list set in ENV).
+// * the page content is uncached
+// * the cache errors
+export const pageCacheMiddleware = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  if (!PAGE_CACHE_ENABLED) return next()
+
+  const pageType = req.path.split("/")[1]
+  if (!cacheablePath(pageType)) return next()
+
+  // @ts-ignore
+  const hasUser = !!req.user
+  if (hasUser) return next()
+
+  const handleCacheGet = (err, html) => {
+    if (err) return next()
+
+    if (html) return res.send(html)
+
+    return next()
+  }
+
+  res.on("finish", () => {
+    if (res.locals.PAGE_CACHE) {
+      cacheHtmlForPage({ ...res.locals.PAGE_CACHE })
+    }
+  })
+  cache.get(cacheKey(req.url), handleCacheGet)
+}
+
+// `key` should be a full URL w/ query params, and not a path.
+// This is to separate URL's like `/collect` and `/collect?acquireable=true`.
+export const cacheKey = (key: string) => {
+  return PAGE_CACHE_NAMESPACE + "-" + PAGE_CACHE_VERSION + key
+}
+
+const cacheablePath = (pageType: string) => {
+  return cacheablePageTypes.includes(pageType)
+}
+
+// `key` is a full URL w/ possibly query params
+export const cacheHtmlForPage = ({ status, key, html }) => {
+  if (!PAGE_CACHE_ENABLED || status !== 200) return
+
+  const pageTypeWithQueryParams = key.split("/")[1]
+  const pageType = pageTypeWithQueryParams.split("?")[0]
```

Comment:
There are 2 separate implementations for extracting a "type," one above from a request object, and this one from a URL string. If any differences creep in between these, it could lead to strange and undetected bugs. Maybe this method could be updated to accept a request object instead of the request's `.url` property, and then they could be made to match exactly.

Reply:
Good call, will look into that!